File: src/main/java/com/refinedmods/refinedstorage/network/tiledata/TileDataParameterMessage.java
package com.refinedmods.refinedstorage.network.tiledata;

import com.refinedmods.refinedstorage.tile.data.TileDataManager;
import com.refinedmods.refinedstorage.tile.data.TileDataParameter;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraftforge.network.NetworkEvent;

import java.util.function.Supplier;

public class TileDataParameterMessage {
    private final BlockEntity tile;
    private final TileDataParameter parameter;
    private final boolean initial;

    public TileDataParameterMessage(BlockEntity tile, TileDataParameter parameter, boolean initial) {
        this.tile = tile;
        this.parameter = parameter;
        this.initial = initial;
    }

    public static TileDataParameterMessage decode(FriendlyByteBuf buf) {
        int id = buf.readInt();
        boolean initial = buf.readBoolean();

        TileDataParameter parameter = TileDataManager.getParameter(id);

        if (parameter != null) {
            try {
                parameter.setValue(initial, parameter.getSerializer().read(buf));
            } catch (Exception e) {
                // NO OP
            }
        }

        return new TileDataParameterMessage(null, null, initial);
    }

    public static void encode(TileDataParameterMessage message, FriendlyByteBuf buf) {
        buf.writeInt(message.parameter.getId());
        buf.writeBoolean(message.initial);

        message.parameter.getSerializer().write(buf, message.parameter.getValueProducer().apply(message.tile));
    }

    public static void handle(Supplier<NetworkEvent.Context> ctx) {
        ctx.get().setPacketHandled(true);
    }
}

File: src/main/kotlin/com/refinedmods/refinedstorage/network/tiledata/TileDataParameterMessage.kt
package com.refinedmods.refinedstorage.network.tiledata

import com.refinedmods.refinedstorage.tile.data.TileDataManager
import com.refinedmods.refinedstorage.tile.data.TileDataParameter
import net.minecraft.block.entity.BlockEntity
import net.minecraft.network.PacketByteBuf
import java.util.function.Supplier

class TileDataParameterMessage<T: Any?, E: BlockEntity?>(
        private val tile: E?,
        private val parameter: TileDataParameter<T?, E?>?,
        private val initial: Boolean
) {

    companion object {
        fun <T: Any?, E: BlockEntity?> decode(buf: PacketByteBuf): TileDataParameterMessage<T?, E?> {
            val id: Int = buf.readInt()
            val initial: Boolean = buf.readBoolean()
            val parameter = TileDataManager.getParameter<T, E>(id)
            if (parameter != null) {
                try {
                    parameter.setValue(initial, parameter.serializer.read(buf))
                } catch (e: Exception) {
                    // NO OP
                }
            }
            return TileDataParameterMessage<T?, E?>(null, null, initial)
        }

        fun <T: Any?, E: BlockEntity?> encode(message: TileDataParameterMessage<T?, E?>, buf: PacketByteBuf) {
            buf.writeInt(message.parameter!!.id)
            buf.writeBoolean(message.initial)
            message.parameter.serializer.write(buf, message.parameter.valueProducer.apply(message.tile))
        }

//        fun <T: Any?, E: BlockEntity?> handle(message: TileDataParameterMessage<T?, E?>?, ctx: Supplier<NetworkEvent.Context>) {
//            ctx.get().setPacketHandled(true)
//        }
    }

}